### ISSUE TITLE
UserUtteranceReverted event incorrectly replays actions

### ISSUE THREAD

**Issue by jctrouble** [RasaHQ/rasa_core#193]
**Rasa Core version**:
master (5a0f0ea445da8f6cbfe53402aa70b1900fd23f66)

**Python version**: 
3.6.3

**Operating system** (windows, osx, ...):
OSX 10.13.2

## Example
Sample project illustrating the error:
https://github.com/twineai/rasa_core-1/tree/rewind_error/examples/rewindbot

![screen shot 2018-01-25 at 12 51 43 pm](https://user-images.githubusercontent.com/1136226/35411703-8e627c7e-01ce-11e8-875a-3d27f26ce240.png)

## Issue
In the example above, the `do_inconsequential_thing` action issues a `UserUtteranceReverted` event. After the event is applied, the bot predicts the next actions to be the actions that were previously ran, whereas I'd expect the bot's state to be simply as if the utterance never occurred at all.

## Diagnosis
I believe it's the result of the following:
*Prior to utterance*
```
Tracker Event History =
   - ActionExecuted(Listen)
   - UserUttered(do_destructive_thing)
   - ActionExecuted(do_destructive_thing_one)
   - ActionExecuted(do_destructive_thing_two)
```

*At this moment, the bot is in the process of listening*
*Now the utterance comes in, and the next action runs (the one which yields the `UserUtteranceReverted`)*
```
Tracker Event History =
   - ActionExecuted(Listen)
   - UserUttered(do_destructive_thing)
   - ActionExecuted(do_destructive_thing_one)
   - ActionExecuted(do_destructive_thing_two)
   - ActionExecuted(listen)
   - UserUttered(do_inconsequential_thing)
   - ActionExecuted(do_inconsequential_thing)
```

*The `UserUtteranceReverted` then rewinds the second `UserUttered`, yielding a state like this:*
```
Tracker Event History =
   - ActionExecuted(Listen)
   - UserUttered(do_destructive_thing)
   - ActionExecuted(do_destructive_thing_one)
   - ActionExecuted(do_destructive_thing_two)
   - ActionExecuted(listen)
```

*Then the policy looks at the state and says "what do I do in this situation", the model sees the last **completed** action (this is important) being the listen, and tries to recommend an action. This queries the last message for the intent (`do_destructive_thing` in this case), and then runs the associated actions.*

I believe the `UserUtteranceReverted` should also rewind the `ActionExecuted(listen)` that occurs immediately prior to the user utterance, so that the model always predicts the next action as being `listen`, thus making the bot behave as if the utterance never occurred (barring side effects from the actions, of course).

## Config

**Content of domain file**:
```yaml
intents:
  - do_destructive_thing
  - do_inconsequential_thing

actions:
  - actions.ActionDoInconsequentialThing
  - actions.ActionDoDestructiveThingOne
  - actions.ActionDoDestructiveThingTwo
```

**Content of stories**:
```
## greet
* do_destructive_thing
    - action_do_destructive_thing_one
    - action_do_destructive_thing_two

## do thing
* do_inconsequential_thing
    - action_do_inconsequential_thing
```

**Content of actions**:
```python
class ActionDoDestructiveThingOne(Action):
    def name(self):
        return 'action_do_destructive_thing_one'

    def run(self, dispatcher, tracker, domain):
        dispatcher.utter_message("Doing destructive thing one")
        return []


class ActionDoDestructiveThingTwo(Action):
    def name(self):
        return 'action_do_destructive_thing_two'

    def run(self, dispatcher, tracker, domain):
        dispatcher.utter_message("Doing destructive thing two")
        return []


class ActionDoInconsequentialThing(Action):
    def name(self):
        return 'action_do_inconsequential_thing'

    def run(self, dispatcher, tracker, domain):
        dispatcher.utter_message("Doing an inconsequential thing that shouldn't be remembered")
        return [UserUtteranceReverted()]
```

**Comment by Timsey** [comment]
I ran into the same issue a while back and fixed it by returning [UserUtteranceReverted(), ActionReverted()] instead of just [UserUtteranceReverted()] in any action that reverts a user utterance. This is not very clean though. Is there a reason for the current implementation?

**Comment by tmbo** [comment]
@jctrouble Thanks, this is a very good bug report :+1: 

I totally get your point, and agree that this is not a good solution. I also saw the PR - but I was wondering if it might not make more sense to keep the reverting as it is right now but then instead of prediction another action notice that the current action is action listen which means that the bot should not predict anything. Basically the check if the bot should be listening needs to be moved after the point of applying the current events. Does that make sense? What do you think?

**Comment by jctrouble** [comment]
@tmbo I think I follow, and it seems reasonable on the surface, but something doesn't quite feel right. To pressure test, let's look at the following situation (I'm using a remote action factory here to cleanly illustrate):

**Remote**: Here's a message (`foo`)
*Bot*: Run this action (`foo`)
**Remote**: Done running`foo`, no events to report
*Bot*: Here's the next action (`listen`)
**Remote**: Got this new message while listening (`bar`)
*Bot*: Run this action (`bar`)
**Remote**: Done running`bar`, here's a `UserUtteranceReverted`

At this moment, the bot (still needing to tell the remote what to do next) would consult a method like `_get_next_action` which ultimately hits the policy to decide what to do. Here seems like a reasonable place to say `if tracker.latest_action_name == ACTION_LISTEN_NAME: return ActionListen()` (rather than hitting up the policy).

This would hand the correct action back to the remote and it would start listening. It also will have the effect of ultimately calling `_log_action_on_tracker`, recording another listen action in the event history, yielding:

```
Tracker Event History =
   - ActionExecuted(Listen)
   - UserUttered(foo)
   - ActionExecuted(foo)
   - ActionExecuted(listen)
   - ActionExecuted(listen)
```

Imagine now that the conversation ends up doing `bar` 3 more times, the event history will look something like:

```
Tracker Event History =
   - ActionExecuted(Listen)
   - UserUttered(foo)
   - ActionExecuted(foo)
   - ActionExecuted(listen)
   - ActionExecuted(listen)
   - ActionExecuted(listen)
   - ActionExecuted(listen)
   - ActionExecuted(listen)
```

While the behavior would be correct, this outcome doesn't really pass the smell test for me for a couple of reasons:
- In my mind, having the last action in the tracker be an `ActionExecuted(listen)` is an indication that the current state of the bot is listening/waiting for a new message. Having the data structure accurately reflect the current state of the conversation seems valuable. As a follow on to that, having a stacking set of `ActionExecuted(listen)` seems odd, and would make debuggers do a double take. Not to mention that it would clog up the history and potentially impact future predictions.
- I also believe that calling `_get_next_action` whilst listening is an error/warning situation (pretend we aren't talking about this revert situation – it seems like incorrect usage regardless).

However, in this remote example above, we still need to respond to the caller with the next action to run after applying the `UserUtteranceReverted`, which necessarily needs to call `_get_next_action`, which (if you believe my assertions above) is bad.

---

Even looking at other places to do the event inspection (i.e. `_predict_next_and_return_state`, `should_predict_another_action`), it still seems to me a little bit like papering over bad data.

Thoughts?

**Comment by tmbo** [comment]
Yes, that is very valid, thanks a lot for the very detailed examples.

**Comment by tmbo** [comment]
I think we resolved that as part of #194.
